### PR TITLE
Text generation refactor and RETRO text generation implementation

### PR DESCRIPTION
**Pull request by yidong72**
# What does this PR do ?

Refactor the text generation code to be model independent by using different inference strategies.

Added the GPT inference strategy and prompt learning inference strategy and verify they are working properly.

Implement the RETRIVAL service that can provides retrieval results on the fly.

Implement the RETRO model text generation and fix a default bias problem.

### CODE REVIEW

**Comment by okuchaiev** [comment]
@vadam5  can you please review this with respect to GPT
@MaximumEntropy can you please review this with respect to T5?

**Review by MaximumEntropy — COMMENTED** [review]
Minor comments.

File: examples/nlp/language_modeling/conf/megatron_retro_inference.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  accelerator: gpu
+  logger: False # logger provided by exp_manager
+  precision: 16 # 16, 32, or bf16
+
+inference_batch_size: 2
+tensor_model_parallel_size: 1
+pipeline_model_parallel_size: 1
+pipeline_model_parallel_split_rank: 0 # used for encoder and decoder model
+retro_model_file: null  # RETRO nemo file path
+prompts: # prompts for GPT inference
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_retro_inference.yaml:26]: you are right. fixed

File: examples/nlp/language_modeling/megatron_retro_eval.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    if model_path:
+        save_restore_connector.model_extracted_dir = model_path
+
+    model_cfg = MegatronRetrievalModel.restore_from(
+        model_path, trainer=trainer, return_config=True, save_restore_connector=save_restore_connector,
+    )
+
+    with open_dict(model_cfg):
+        # work around for the fused softmax bug
+        model_cfg.masked_softmax_fusion = False
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_retro_eval.py:65]: good catch. fixed

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py [hunk 3]
```diff
@@ -342,11 +354,10 @@ def test_step(self, batch, batch_idx):
         return self.validation_step(batch, batch_idx)
 
     def test_epoch_end(self, outputs):
-        averaged_loss = average_losses_across_data_parallel_group(outputs)
-        logging.info(f'test_loss: {averaged_loss[0]}')
-        self.log(
-            'consumed_samples', self.compute_consumed_samples(self.trainer.global_step - self.init_global_step),
-        )
+        averaged_loss = torch.stack(outputs).mean()
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:357]: because in the validation step, the returned loss is already averaged across the data parallel_group. This is redudant.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 4]
```diff
@@ -950,7 +950,7 @@ def forward(
             hidden_states.shape[0],
             hidden_states.shape[2],
         )
-        empty_bias = torch.zeros(dim, dtype=hidden_states.dtype, device=hidden_states.device)
+        default_biase = self.cross_attention.dense.bias
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:953]: fixed

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 5]
```diff
@@ -962,7 +962,7 @@ def forward(
             chunk_id = self.current_len // chunk_size
             if chunk_id <= 0:
                 # if sequence length less than chunk size, do an early return
-                return torch.zeros_like(hidden_states), empty_bias
+                return torch.zeros_like(hidden_states), default_biase
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:965]: fixed

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 6]
```diff
@@ -978,7 +978,7 @@ def forward(
 
         # if sequence length less than chunk size, do an early return
         if n < self.chunk_size and set_inference_key_value_memory and inference_max_sequence_len is not None:
-            return torch.zeros_like(hidden_states), empty_bias
+            return torch.zeros_like(hidden_states), default_biase
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:981]: fixed

File: nemo/collections/nlp/modules/common/text_generation_utils.py [hunk 7]
```diff
@@ -331,26 +300,23 @@ def synced_generate(
 ):
     context_length = context_length_tensor.min().item()
     tokenizer = model.tokenizer
-    tokens, attention_mask, position_ids = get_batch(model, tokenizer, context_tokens_tensor)
+    # tokens, attention_mask, position_ids = get_batch(model, tokenizer, context_tokens_tensor)
```
Comment (vadam5) [inline nemo/collections/nlp/modules/common/text_generation_utils.py:303]: This comment can be deleted

**Review by vadam5 — CHANGES_REQUESTED** [review]
Thanks for the refactor of prompt learning/gpt model inference code! Overall looks good, just left a few comments.

File: examples/nlp/language_modeling/megatron_retro_eval.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+from nemo.core.config import hydra_runner
+
+
+"""
+This is the script to run RETRO Model text generation.
+
+Usage:
+    Assume the model has TP=1, PP=1
+    run greedy inference from a nemo file:
+        python megatron_gpt_eval.py \
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_retro_eval.py:34]: Good catch. fixed.

File: examples/nlp/language_modeling/conf/megatron_retro_inference.yaml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+retro_model_file: null  # RETRO nemo file path
+
+use_predict_method: False  # whether to use the predict method
+
+prompts: # prompts for RETRO model inference
+  - "hello,"
+  - "good morning,"
+  - "good afternoon,"
+  - "good evening,"
+server: False  # whether launch the inference server
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_retro_inference.yaml:34]: agree. I will remove it

File: nemo/collections/nlp/modules/common/text_generation_utils.py (hunk 7 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/text_generation_utils.py:303]: removed

**Review by vadam5 — APPROVED** [review]
LGTM! Thank you
